rules/manager.go

```diff
@@ -589,8 +589,18 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
 			)
 
 			app := g.opts.Appendable.Appender()
+			var appErr error
 			seriesReturned := make(map[string]labels.Labels, len(g.seriesInPreviousEval[i]))
 			defer func() {
+				if appErr != nil {
+					app.Rollback()
```

Comment:
We are also calling Commit() below. We should do only 1 of commit or rollback (especially after https://github.com/prometheus/prometheus/pull/7593). And I feel we should be committing the samples which did not give any error instead of rolling back everything.

Reply:
You are right, we should not commit.

However, I think we should roll back everything, just like we do for a failed scrape.

Reply:
Sorry: the commit was right: If we roll back, we commit staleness markers. We take a new appender here.

This is similar to what we do in #7593.